### DISCUSSION TITLE
Attach new global string validation functions directly to z.string()

### DISCUSSION THREAD

**Discussion by jgoux** [colinhacks/zod#366] [Q&A] [answered]
Hello all,

Is it possible to extend z.string() to add other string validation functions in the vein as the built-in uuid and email functions? Something like `z.string().iban()`.

This would be very handy for validations that are common across the codebase and to handle i18n on the error message. I know I could use refinement but it's not convenient for often used validation rules.

I made a special ErrorMap for i18next and it would automatically be handled by this case : https://gist.github.com/jgoux/720f011a4428e64f50bebc9938219518#file-zod-ts-L60-L68

I'm currently using the v3 in alpha. 😄

**Comment by colinhacks** [comment, marked as answer]
The best part (in my opinion) of v3 is that anyone can define their own schema classes by simply subclassing `ZodType`. In your case, you could directly subclass ZodString and add these methods yourself. It wouldn't be super difficult, just look at the ZodString implementation to see how it works. 

Besides `iban` what methods are you looking to add? I might be amenable to adding them to core. I'm a little surprised I don't get more requests for other string validation methods.

Alternatively consider using the "re-export pattern". 

```ts
// customzod.ts;
export * from 'zod';
export const iban = ()=>z.string().refine(validateIban, { message: "Invalid IBAN" });
```

```
// server.ts
import * from "./customzod";

z.object({
  iban: z.iban()
})
````

That custom error map looks really great! If you think that would be useful for other people, consider publishing it to NPM and I'll link to it on the README.

**Reply by jgoux** [reply]
I like the re-export pattern, I'm already using it to customize the error map! 👍 

I have another little issue, I'm trying to share my validation schema between my frontend and my backend. Problem is that I don't have to exact same "z" instances on both environment. Basically my frontend "z" is my backend "z" + setErrorMap : 

```ts
// frontend code
import { z } from "@clovis/server/src/app/config/zod";

import { i18n } from "../i18next";
import { createCustomErrorMap } from "./create-custom-error-map";

const customErrorMap = createCustomErrorMap(i18n);
z.setErrorMap(customErrorMap);

export { z };

```

but my schemas are defined on the backend using the "z" backend instance (I don't have i18n on the backend) : 

```ts
// backend code
import { z } from "../../../../config/zod";

export const inputValidation = z.object({
  name: z.string().max(10),
});
```

Then on the frontend I import the schema and pass it to my useForm hook : 
```ts
// frontend code
import { inputValidation } from "@clovis/server/src/app/domain/project/task/create-task/create-task.validation";

// ... code
useForm(inputValidation); // <- needs to be tied to the frontend "z" instance or it won't have i18n support
```

Do you have any idea how I could use `inputValidation` but with the "z" instance I export in the first code sample? The sane way would be to make a factory that accept "z" for each schema but it's not convenient to use and would impact the backend code as well.

Maybe we could pass the error map later as an option to `parse`?

EDIT : 

Finally I prefer the less magic approach and use a factory : 
```ts
import type { Zod } from "../../../../config/zod";
import { z as zod } from "../../../../config/zod";

// By passing the server instance by default I can just createSchema() on the backend
export const createSchema = (z: Zod = zod) =>
  z.object({
    name: z.string().max(10),
  });
```

But if you have a good idea about this subject I'm listening. 😄

**Reply by colinhacks** [reply]
Actually until very recently the *only* way to use error maps was an a parameter to parse:


```ts
z.string().parse('asdf', { errorMap: myCustomErrorMap })
```


You're using a "weak version" of the re-export pattern (to great effect!). The approach I described is different, in that it lets you add additional methods to the `z` object (e.g. `z.iban()`). This requires using the `export * from` approach. 

As for the multiple `z` values...there's lots of ways to approach this. The factory is one. Or find some way to share a single `z` between client and server. Not really sure what's best here.

Unrelated note, but I usually define my schemas on the client and import them into my server - less chance of any credential leaks, etc.

**Reply by jgoux** [reply]
I didn't realized you could pass ErrorMap to parse! This is awesome and totally solve my double z issue! 💯💯💯

Noted for the re-export pattern, I'll use the "string" version instead which is more like I wanted.

Thanks for the tips Colin. ☺️

**Reply by colinhacks** [reply]
You bet!

**Reply by DominikSerafin** [reply]
@colinhacks I've read multiple issues, PRs and discussions here in this repo, and somehow I still can't come up with a solution to simply create a new custom validation method for primitives that is chainable (in Zod v3).

What I'm basically trying to achieve is to be able to write something like this:

```typescript
import * as z from './zod-extended';

const schema = z.object({
  text: z.string().min(5).nonewlines().max(200),
});
const parsed = schema.parse({text: 'example'});
```

This example has `nonewlines()` method chained between other string methods (which should throw an error if string has any newline characters). 

This is just an example, though, as I'd want to find a way to add a custom validation methods (with custom error code & message) for some of the other primitive types as well (string, number, bigint, date).

The re-export pattern doesn't let me chain my custom method. And I've tried the subclass approach, but I wasn't able to figure out how to do that properly (I've looked into `ZodString` implementation, like you've suggested).

I would really appreciate your help.

**Reply by colinhacks** [reply]
Here's a simple example of implementing a subclass of ZodString. As long as all your methods delegate to `.refine` and `.transform`, this is fairly straightforward. 

```ts
class CustomString extends z.ZodString {
  toNumber() {
    return this.transform((val) => val.length);
  }

  static create = (): CustomString => {
    return new CustomString({
      checks: [],
    });
  };
}

const mySchema = CustomString.create().toNumber();
```

If you want to modify the parsing logic, add additional state/properties, etc, things get more complicated. You'll just have to bite the bullet and learn how the existing schema types work.

**Reply by DominikSerafin** [reply]
@colinhacks thanks for the reply! I've moved this discussion to https://github.com/colinhacks/zod/discussions/516 (so it's a bit cleaner)
